EKF/EKFGSF_yaw_estimator.cpp

```diff
@@ -0,0 +1,827 @@
+	_do_emergency_yaw_reset = false;
+	_ekf_gsf_yaw_variance = sq(M_PI_2_F);
+	memset(&_ekf_gsf, 0, sizeof(_ekf_gsf));
+	const float yaw_increment = 2.0f * M_PI_F / (float)N_MODELS_EKFGSF;
+	for (uint8_t model_index = 0; model_index < N_MODELS_EKFGSF; model_index++) {
+		// evenly space initial yaw estimates in the region between +-Pi
+		_ekf_gsf[model_index].X[2] = -M_PI_F + (0.5f * yaw_increment) + ((float)model_index * yaw_increment);
+
+		// All filter models start with the same weight
+		_ekf_gsf[model_index].W = 1.0f / (float)N_MODELS_EKFGSF;
+
+		if (_control_status.flags.gps) {
+			_ekf_gsf[model_index].X[0] = _gps_sample_delayed.vel(0);
+			_ekf_gsf[model_index].X[1] = _gps_sample_delayed.vel(1);
+			_ekf_gsf[model_index].P[0][0] = sq(_gps_sample_delayed.sacc);
+			_ekf_gsf[model_index].P[1][1] = _ekf_gsf[model_index].P[0][0];
+		} else {
+			// no GPS so assume we are in a semi-static condition and moving no more that 0.5 m/s
+			_ekf_gsf[model_index].P[0][0] = sq(0.5f);
+			_ekf_gsf[model_index].P[1][1] = sq(0.5f);
+		}
+
+		// use half yaw interval for yaw uncertainty
+		_ekf_gsf[model_index].P[2][2] = sq(0.5f * yaw_increment);
+	}
+}
+
+void Ekf::runEKFGSF()
+{
+	// Iniitialise states first time
+	_ahrs_accel = _imu_sample_delayed.delta_vel / fmaxf(_imu_sample_delayed.delta_vel_dt, FILTER_UPDATE_PERIOD_S / 4);
+	if (!_ahrs_ekf_gsf_tilt_aligned) {
+		// check for excessive acceleration to reduce likelihood of large inital roll/pitch errors
+		// due to vehicle movement
+		const float accel_norm_sq = _ahrs_accel.norm_squared();
+		const float upper_accel_limit = CONSTANTS_ONE_G * 1.1f;
+		const float lower_accel_limit = CONSTANTS_ONE_G * 0.9f;
+		const bool ok_to_align = ((accel_norm_sq > lower_accel_limit * lower_accel_limit &&
+			  accel_norm_sq < upper_accel_limit * upper_accel_limit));
+		if (ok_to_align) {
+			initialiseEKFGSF();
+			alignQuatEKFGSF();
+			_ahrs_ekf_gsf_tilt_aligned = true;
+		}
+		return;
+	}
+
+	// calculate common values used by the AHRS prediction models
+	_ahrs_accel_norm = _ahrs_accel.norm();
+	_ahrs_turn_comp_enabled = _control_status.flags.fixed_wing && _params.EKFGSF_tas_default > FLT_EPSILON;
+	if (_ahrs_accel_norm > CONSTANTS_ONE_G) {
+		if (_ahrs_turn_comp_enabled && _ahrs_accel_norm <= 2.0f * CONSTANTS_ONE_G) {
+			_ahrs_accel_fusion_gain = _params.EKFGSF_tilt_gain * sq(1.0f - (_ahrs_accel_norm - CONSTANTS_ONE_G)/CONSTANTS_ONE_G);
+		} else if (_ahrs_accel_norm <= 1.5f * CONSTANTS_ONE_G) {
+			_ahrs_accel_fusion_gain = _params.EKFGSF_tilt_gain * sq(1.0f - 2.0f * (_ahrs_accel_norm - CONSTANTS_ONE_G)/CONSTANTS_ONE_G);
+		}
+	} else if (_ahrs_accel_norm > 0.5f * CONSTANTS_ONE_G) {
+		_ahrs_accel_fusion_gain = _params.EKFGSF_tilt_gain * sq(1.0f + 2.0f * (_ahrs_accel_norm - CONSTANTS_ONE_G)/CONSTANTS_ONE_G);
+	}
+
+	// AHRS prediction cycle for each model - this always runs
+	for (uint8_t model_index = 0; model_index < N_MODELS_EKFGSF; model_index ++) {
+		statePredictEKFGSF(model_index);
+	}
```

Comment:
```c++
	_ahrs_accel_norm = _ahrs_accel.norm();
	_ahrs_turn_comp_enabled = _control_status.flags.fixed_wing && _params.EKFGSF_tas_default > FLT_EPSILON;
	computeAhrsFusionGain()

	// AHRS prediction cycle for each model - this always runs
	for (uint8_t model_index = 0; model_index < N_MODELS_EKFGSF; model_index ++) {
		statePredictEKFGSF(model_index);
	}
```
```c++
void Ekf::computeAhrsFusionGain()
{
   	// calculate common values used by the AHRS prediction models
   	// see https://www.desmos.com/calculator/c8hztzyrni
	float attenuation = 2.f;
	if (_ahrs_turn_comp_enabled 
	    && _ahrs_accel_norm > CONSTANTS_ONE_G
	    && _ahrs_accel_norm <= 2.f * CONSTANTS_ONE_G) {
		attenuation = 1.f;
	}

	const float delta_accel = _ahrs_accel_norm - CONSTANTS_ONE_G;
	const float accel_ratio_abs = math::min(fabsf(delta_accel), 1.f / attenuation) / CONSTANTS_ONE_G;
	_ahrs_accel_fusion_gain = _params.EKFGSF_tilt_gain * sq(1.f - attenuation * accel_ratio_abs);
}

Reply:
Comparison with the other implementation : https://www.desmos.com/calculator/ycph8sadhx
Also note that in the previous implementation, when `_ahrs_accel_norm < 0.5 g` or ` _ahrs_accel_norm > 1.5 g`, the `_ahrs_accel_fusion_gain` isn't updated, so a spike can be fused with a high gain.

Reply:
We need to be able to do tilt correction with 45 degree banked turns + some gust margin when operating as a fixed wing. These settings are working OK for replay of the logs we have, but we may end up adjusting the logic as we get more data.

Reply:
@priseborough To show you that the asymmetric curve is also handled: https://www.desmos.com/calculator/kq95hay45h
![DeepinScreenshot_select-area_20200220135708](https://user-images.githubusercontent.com/14822839/74935644-e0f6a500-53e8-11ea-8f29-8bac149e4298.png)

Reply:
I've fixed the bug that can leave the gain at the previous value and have moved the gain calculation into a function. See https://github.com/PX4/ecl/pull/751/commits/340df697ed2e8ba3d9f2aa289dc1dd3d4d462a2e